lib/azure/armrest/configuration.rb

```
@@ -0,0 +1,134 @@
+      #
+      # Example:
+      #
+      #   config = Azure::Armrest::Configuration.new(
+      #     :client_id  => 'xxxx',
+      #     :client_key => 'yyyy',
+      #     :tenant_id  => 'zzzz'
+      #   )
+      #
+      def initialize(hash)
+        hash.each{ |key,value| send("#{key}=", value) }
+
+        unless client_id && client_key
+          raise ArgumentError, "client_id and client_key must both be specified"
+        end
+
+        # Set defaults if not already set
+        @accept       ||= 'application/json'
+        @content_type ||= 'application/json'
+        @grant_type   ||= 'client_credentials'
+        @api_version  ||= '2015-01-01'
+      end
+
+      # Returns the tokens class variable hash.
+      def tokens
+        @@tokens
+      end
+
+      # A combination of grant_type, tenant_id, client_id and client_key,
+      # as a single string joined by underscores.
+      #
+      # Used to identify unique sessions.
+      #
+      def as_cache_key
+        "#{grant_type}_#{tenant_id}_#{client_id}_#{client_key}"
+      end
+
+      alias cache_key as_cache_key
+
+      # Returns the token for the given cache_key, or sets it if it does not
+      # exist or it has expired.
+      #
+      def token
+        @token, @token_expiration = @@tokens[cache_key] if @token.nil?
+
+        if @token.nil? || Time.now > @token_expiration
+          @token, @token_expiration = fetch_token
+        end
+
+        @token
+      end
+
+      # Set the token value. Setting it in this manner will automatically
+      # set the expiration time for 1 hour. To explicitly set the expiration
+      # time, use the :set_token method instead.
+      #
+      def token=(value)
+        set_token(value)
+      end
+
+      # Set the token value and expiration time. If no time is specified, then
+      # the time is set for 1 hour.
+      #
+      def set_token(token, token_expiration = Time.now + 3600)
```

It is Azure that configures the token expiration. Setting a default expiration has no meaning and misleads.